lib/db.js

```diff
@@ -92,22 +93,30 @@ class LimitDBRedis extends EventEmitter {
     this.buckets[key] = utils.buildBucket(bucket);
   }
 
-  bucketKeyConfig(type, key) {
-    const fromCache = type.overridesCache.get(key);
+  /**
+   * @param {string} type
+   * @param {object} params
+   * @returns
+   */
+  bucketKeyConfig(type, params) {
+    if (typeof params.configOverride === 'object') {
+      return utils.normalizeType(params.configOverride);
```

Comment:
Notice same code path that adds `ttl` will be used ✅

Reply:
`normalizeType()` will also set the `overrides` and `overridesCache` attributes on `params.configOverride`, which does not make much sense imo. In particular, creating an instance of `lru-cache` on each call to `take()` for example might be costly? I think it would be preferable to split `normalizeType()` into two smaller methods.

The other aspect that I dislike about using `normalizeType()` as-is, is that the method mutates the input `params.configOverride` (passed from outside the lib). This is just something that I've seen cause problems in the past, especially in the case when `params.configOverride` could be a reference to an in-memory cache entry that we re-use across multiple calls to `take()` 😬

Reply:
> normalizeType() will also set the overrides and overridesCache attributes on params.configOverride, which does not make much sense imo

100% agree, I think we're just talking about how aggressively we want to fix existing bugs 🤷‍♂️ .  For instance, this is a sample generated `bucketKeyConfig` in limitd-redis today:

```
{
  "size": 10,
  "interval": 1000,
  "per_interval": 5,
  "ttl": 2,
  "overrides": [
    {
      "interval": 1000,
      "per_interval": 100,
      "size": 100,
      "ttl": 1,
      "overrides": [],
      "overridesCache": {},
      "name": "127.0.0.1"
    },
    {
      "interval": 1000,
      "per_interval": 50,
      "size": 50,
      "ttl": 1,
      "overrides": [],
      "overridesCache": {},
      "name": "local-lan",
      "match": {}
    },
    {
      "size": 1,
      "interval": 3600000,
      "per_interval": 2,
      "ttl": 1800,
      "overrides": [],
      "overridesCache": {},
      "name": "10.0.0.1"
    },
  ],
  "overridesCache": {}
}
``` 

Notice that each of the overrides ALSO has an overrides cache, whether they contain a `match` regexp or not.  What's the difference between this:
```
const fromCache = type.overridesCache.get(params.key)
```
And simply referencing a bucket config by name?
```
const theThing = type.overrides[paramskey]
```

Seems as if the cache was intended for use by base bucket types when regexes were in play for override matching, but instead we've started:

1. Using the cache for non-regex entities, effectively bloating the cache with entities we can simply reference quickly by key anyway.
2. Creating cache entities on all overrides.  For instance, `{"per_second":100}` goes through `normalizeType` and becomes `{"interval":1000,"per_interval":100,"size":100,"ttl":1,"overrides":[],"overridesCache":{}}`.  This overrides cache will **never** be used or referenced.  Since we recursively [normalizeType](https://github.com/auth0/limitd-redis/blob/2b44821a2f6bc0ad80a778310367483a0d2f7d9f/lib/utils.js#L37), then [set a truthy default for type.overrides](https://github.com/auth0/limitd-redis/blob/2b44821a2f6bc0ad80a778310367483a0d2f7d9f/lib/utils.js#L36), we [always get the LRU cache](https://github.com/auth0/limitd-redis/blob/2b44821a2f6bc0ad80a778310367483a0d2f7d9f/lib/utils.js#L51) 🤦‍♂️ .

So know that I'm seeing these things... but was intentionally trying to keep the scope narrow.  How aggressively do we want to fix the flawed type normalization?  What kind of risk/time footprint are we willing to take on here?

Other problems I have with this code are widespread and varied.  Here are a couple examples:

- Validation returning an error instead of throwing to a non-callback control flow.  We're touching this as well... do we alter the validation logic control flow as part of adding override parameters?
- PUT (lua script) does not take into account dripped tokens in the interval between the last REDIS write and its invocation.  This means that if PUT is called a significant amount of time after TAKE, then the incorrect number of tokens will be in the bucket.  Probably not going to be a widespread problem given the shape of our usage (in which PUT is often called in quick succession), but a footgun waiting to go off nonetheless...

So.... help me with scoping here, lest this become another month of May/June in which we fix all the things and it hinders our ability to deliver.

Reply:
Wow, it is even more flawed than what I thought 😬 

> How aggressively do we want to fix the flawed type normalization?

My take: we don't touch anything about the static config, we just make sure the new `configOverride` does not inherit any of those flaws at the outset. Just extracting part of the `normalizeType()` method into a smaller reusable method does not seem too risky imo.

I wouldn't touch anything about validation logic and just follow the existing semantics.

Regarding the `PUT` script, it should probably take into account the dripped tokens... but I wouldn't touch that now, just raise this to platform-services team and they should assess this bug.

Reply:
OK - very gently extracted and used a sub-set of the existing `normalizeType` when a configuration override is used such that no LRU cache is created.